modules/percolator/src/main/java/org/elasticsearch/percolator/ExtractQueryTermsService.java

```diff
@@ -153,76 +334,71 @@ public static void extractQueryTerms(Query query, ParseContext.Document document
                         continue;
                     }
 
-                    Set<Term> temp;
+                    Result temp;
                     try {
                         temp = extractQueryTerms(clause.getQuery());
+                        if (temp.noTerms) {
+                            continue;
+                        } else {
+                            verifiedIfMatchAll = false;
+                        }
                     } catch (UnsupportedQueryException e) {
                         uqe = e;
                         continue;
                     }
-                    bestClause = selectTermListWithTheLongestShortestTerm(temp, bestClause);
+                    bestClause = selectTermListWithTheLongestShortestTerm(temp.terms, bestClause);
                 }
                 if (bestClause != null) {
-                    return bestClause;
+                    // 1 required clause, no prohibited clause, maybe some optional clauses, so it is verified:
+                    boolean verified = minimumShouldMatch == 0 && numProhibitedClauses == 0 && numRequiredClauses == 1;
+                    return new Result(verified, bestClause);
                 } else {
                     if (uqe != null) {
+                        // we're unable to select the best clause and an exception occurred, so we bail
                         throw uqe;
+                    } else {
+                        // We didn't find a clause and no exception occurred, so this bq only contained MatchAllDocsQueries
+                        // and / or MatchNoDocsQueries, so we mark this result as no terms, because no terms were extracted:
+                        return new Result(verifiedIfMatchAll, true);
                     }
-                    return Collections.emptySet();
                 }
             } else {
-                Set<Term> terms = new HashSet<>();
+                List<Query> disjunctions = new ArrayList<>(numOptionalClauses);
                 for (BooleanClause clause : clauses) {
-                    if (clause.isProhibited()) {
-                        // we don't need to remember the things that do *not* match...
-                        continue;
+                    if (clause.isScoring()) {
```

Comment:
The reason did work out is that if there were must clauses this else would not have been executed, the above if clause would have been executed. Tests would have failed if we got here and there were must clauses in the bq.

I agree that this looks confusing and we should use `clause.getOccur() == SHOULD` instead.